### PR TITLE
fix(Popup): Prevent closing on content click

### LINKED ISSUE #14800: Popup not properly handles click inside
### Environment Information

- **Package version(s)**: `@fluentui/react-northstar@0.51.0`
- **Browser and OS versions**: N/A

CSB: https://codesandbox.io/s/flamboyant-yalow-8vjey
popup-bug

#### Actual behavior:

1. Hover a button
2. Click on a button
3. Click on green background
4. 🟥 Popup was closed

#### Expected behavior:

4. ✔️ Popup should be still opened

### Priorities and help requested:

Are you willing to submit a PR to fix? Yes
Requested priority: Normal
Products/sites affected: N/A

### PR DESCRIPTION
**Pull request by assuncaocharles**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #14800
- [ ] Include a change request file using `$ yarn change`

#### Description of changes

This PR prevent the popup content to closes if the blur event doesn't is not coming from keyboard. We are listening click events and handling it so it will not after events coming from mouse. The issue is happening because related target will be null since the popup content described doesn't have `tabIndex` so `relatedTarget` for the blue event will always be `null`

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by assuncaocharles** [comment]
@layershifter Can you check if it's proper fix? If so I will add tests

File: packages/fluentui/react-northstar/src/components/Popup/Popup.tsx [hunk 1]
```diff
@@ -370,11 +370,11 @@ export const Popup: React.FC<PopupProps> &
   };
 
   const shouldBlurClose = e => {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Popup/Popup.tsx:372]:
```suggestion
  const shouldBlurClose = (e: TYPE): boolean => {
```

Let's improve typings

File: packages/fluentui/react-northstar/src/components/Popup/Popup.tsx [hunk 2]
```diff
@@ -369,12 +369,13 @@ export const Popup: React.FC<PopupProps> &
...
-      !popupContentRef.current ||
-      (!e.currentTarget.contains(e.relatedTarget) && !popupContentRef.current.contains(e.relatedTarget))
-    );
+  const shouldBlurClose = (e: React.FocusEvent) => {
+    const relatedTarget = e.relatedTarget as Node;
+    const isInsideContent = popupContentRef.current?.contains(relatedTarget);
+    const isInsideTarget = e.currentTarget?.contains(relatedTarget);
+    // When clicking in the popup content that has no tabIndex focus goes to body
+    // We shouldn't close the popup in this case
+    return e.relatedTarget && !(isInsideContent || isInsideTarget);
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Popup/Popup.tsx:378]:
```suggestion
    return relatedTarget && !(isInsideContent || isInsideTarget);
```

File: packages/fluentui/e2e/tests/popupClickHandling-test.ts [hunk 3]
```diff
@@ -20,4 +21,15 @@ describe('Popup - on content click', () => {
     // but popup content itself remains to be opened
     expect(await e2e.exists(popupContent)).toBe(true);
   });
+
+  it('should keep it open on content click', async () => {
+    await e2e.hover(popupTrigger); // opens popup
+    await e2e.clickOn(popupTrigger); // clicks setting focus on trigger ( should keep open )
+    await e2e.clickOn(popupContentContainer); // clicks on the container ( focus goes to body but still open)
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupClickHandling-test.ts:28]:
```suggestion

    await e2e.clickOn(popupContentContainer); // clicks on the container (focus goes to body but still open)
    await e2e.isFocused('body')
```

File: packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts [hunk 4]
```diff
@@ -11,9 +11,12 @@ describe('Popup With Close in the content and hover trigger', () => {
 
   xit('A click on content should not close popup, a click button should close popup', async () => {
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts:12]:
```suggestion
  it('A click on content should not close popup, a click button should close popup', async () => {
```

File: packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts [hunk 5]
```diff
@@ -11,9 +11,12 @@ describe('Popup With Close in the content and hover trigger', () => {
...
     await e2e.hover(popupTrigger); // opens popup
+    await e2e.clickOn(popupTrigger); // opens popup
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts:14]: Let's add this as a separate test

File: packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts [hunk 6]
```diff
@@ -9,12 +9,31 @@ describe('Popup With Close in the content and hover trigger', () => {
...
+    await e2e.clickOn(popupContent);
     expect(await e2e.exists(popupContent)).toBe(true);
+
     await e2e.clickOn(popupClose);
     expect(await e2e.exists(popupContent)).toBe(false);
   });
+
+  it('When opened by hover & click, should be kept it opened on content click', async () => {
+    await e2e.hover(popupTrigger);
+
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts:25]:
```suggestion
```

File: packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts [hunk 7]
```diff
@@ -9,12 +9,31 @@ describe('Popup With Close in the content and hover trigger', () => {
...
+    expect(await e2e.exists(popupContent)).toBe(true);
+
+    await e2e.clickOn(popupTrigger);
+    expect(await e2e.isFocused(popupTrigger)).toBe(true);
+
+    expect(await e2e.exists(popupContent)).toBe(true);
+
+    await e2e.clickOn(popupContent);
+
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts:34]:
```suggestion
```

File: packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts [hunk 8]
```diff
@@ -9,12 +9,29 @@ describe('Popup With Close in the content and hover trigger', () => {
...
     expect(await e2e.exists(popupContent)).toBe(false);
   });
+
+  it('When opened by hover & click, should be kept it opened on content click', async () => {
+    await e2e.hover(popupTrigger);
+    expect(await e2e.exists(popupContent)).toBe(true);
+
+    await e2e.clickOn(popupTrigger);
+    expect(await e2e.isFocused(popupTrigger)).toBe(true);
+
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts:29]:
```suggestion
```

File: packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts [hunk 9]
```diff
@@ -9,12 +9,29 @@ describe('Popup With Close in the content and hover trigger', () => {
...
+    expect(await e2e.exists(popupContent)).toBe(true);
+
+    await e2e.clickOn(popupContent);
+    expect(await e2e.exists(popupContent)).toBe(true);
+
+    await e2e.isFocused('body');
```
Comment (layershifter) [inline packages/fluentui/e2e/tests/popupWithCloseInContent-test.ts:35]:
```suggestion
```

**Review by layershifter — APPROVED** [review]

**Pull request merged**